In [2]:
import pandas as pd
import numpy as np

This is just a short example of the use of the crosstab function. It is essentially a form of pivot table that has turned our columns into the variables from the 'status' column, then aggregated all the different grades within those columns using 'margins=True'. 

In [3]:
df=pd.read_csv('lc_loans.csv')
df.head()

status grade
0   Fully Paid     B
1  Charged Off     C
2   Fully Paid     C
3   Fully Paid     C
4      Current     B

In [4]:
df_cross = pd.crosstab(index=df['grade'], columns=df['status'], margins=True)
df_cross

status  Charged Off  Current  Fully Paid  Late     All
grade                                                 
A              1562    50051       20408   469   72490
B              5302    93852       31160  2056  132370
C              6023    88928       23147  2777  120875
D              5007    53281       13681  2308   74277
E              2842    24639        5949  1374   34804
F              1526     8444        2328   606   12904
G               409     1990         643   199    3241
All           22671   321185       97316  9789  450961

We can by default normalize the values by dividing each value by the sum of the values. However, using normalize='index' will normalize by row. For example, for Charged Off and Grade A the sum would be 1562/72490 which comes from the subtotal of the 'All' row for Grade A. 

In [5]:
df_cross_1 = pd.crosstab(index=df['grade'], columns=df['status'], margins=True, normalize='index')
df_cross_1

status  Charged Off   Current  Fully Paid      Late
grade                                              
A          0.021548  0.690454    0.281528  0.006470
B          0.040054  0.709013    0.235401  0.015532
C          0.049828  0.735702    0.191495  0.022974
D          0.067410  0.717328    0.184189  0.031073
E          0.081657  0.707936    0.170929  0.039478
F          0.118258  0.654371    0.180409  0.046962
G          0.126196  0.614008    0.198396  0.061401
All        0.050273  0.712223    0.215797  0.021707

Both the 'All' column and 'All' row should equal to 1.0 if you have done it correctly. This addition to the normalized dataframe shows the proportion of Grades overall, so Grade A makes up 0.16 of the total of all loans and Grade B makes up 0.29 etc. 

In [6]:
df_cross_1['All'] = df_cross['All']/df_cross['All'][-1]
df_cross_1

status  Charged Off   Current  Fully Paid      Late       All
grade                                                        
A          0.021548  0.690454    0.281528  0.006470  0.160746
B          0.040054  0.709013    0.235401  0.015532  0.293529
C          0.049828  0.735702    0.191495  0.022974  0.268039
D          0.067410  0.717328    0.184189  0.031073  0.164708
E          0.081657  0.707936    0.170929  0.039478  0.077177
F          0.118258  0.654371    0.180409  0.046962  0.028614
G          0.126196  0.614008    0.198396  0.061401  0.007187
All        0.050273  0.712223    0.215797  0.021707  1.000000

Checking that the rows truly add up to 1.0. The sum of the rows equals 2 because it is also adding the +1.0 from the All column, but if we minus that then we end up with 1.0.

In [7]:
print(df_cross_1['All'][0:7].sum())
print(df_cross_1.loc['All'].sum())

1.0
2.0


In [8]:
bridges = pd.read_csv('tx19_bridges_sample.csv')

Below we are comparing substructure to superstructure ratings. We have used normalize = all so the proportions are out of all the bridges. For example, out of all the bridges 0.04% have both a failed superstructure and substructure whereas 14.91% have a Very good superstructure and Good substructure rating. 

In [9]:
sub_sup = pd.crosstab(bridges.Substr_rating, bridges.Superstr_rating, normalize='all', margins=True)
sub_sup = sub_sup.round(4)*100
sub_sup

Superstr_rating  Critical  Excellent  Failed  Failing  Fair   Good  Poor  \
Substr_rating                                                              
Critical             0.00       0.00    0.00     0.00  0.01   0.01  0.01   
Excellent            0.00       0.44    0.00     0.00  0.00   0.01  0.00   
Failed               0.00       0.00    0.04     0.00  0.00   0.00  0.00   
Failing              0.00       0.00    0.00     0.01  0.00   0.00  0.00   
Fair                 0.00       0.00    0.00     0.00  1.31   1.68  0.11   
Good                 0.00       0.22    0.00     0.00  0.73  27.99  0.03   
Poor                 0.00       0.00    0.00     0.00  0.24   0.21  0.08   
Satisfactory         0.00       0.01    0.00     0.00  1.65  13.57  0.16   
Serious              0.00       0.00    0.00     0.00  0.03   0.05  0.01   
Very Good            0.00       0.37    0.00     0.00  0.02   1.74  0.00   
All                  0.01       1.04    0.05     0.01  3.98  45.26  0.41   

Superstr_rating  Satisfactory  Serious  Very Good     All  
Substr_rating                                              
Critical                 0.03     0.00       0.00    0.06  
Excellent                0.00     0.00       0.01    0.46  
Failed                   0.00     0.00       0.00    0.05  
Failing                  0.00     0.00       0.00    0.01  
Fair                     3.66     0.02       0.14    6.93  
Good                     7.77     0.01      14.91   51.67  
Poor                     0.52     0.01       0.01    1.06  
Satisfactory            14.32     0.02       1.90   31.63  
Serious                  0.08     0.00       0.00    0.17  
Very Good                0.12     0.00       5.71    7.97  
All                     26.51     0.06      22.68  100.00

Here we have done the same as above but used normalize = columns. This is normalizing based on the superstructure rating, so for failing superstructures 66.67% also have a failing substructure and 33.33% have a satisfactory substructure.

In [10]:
sub_sup2 = pd.crosstab(bridges.Substr_rating, bridges.Superstr_rating, normalize='columns')
sub_sup2 = sub_sup2.round(4)*100
sub_sup2

Superstr_rating  Critical  Excellent  Failed  Failing   Fair   Good   Poor  \
Substr_rating                                                                
Critical             25.0       0.00    6.25     0.00   0.15   0.02   1.44   
Excellent             0.0      42.82    0.00     0.00   0.00   0.02   0.00   
Failed                0.0       0.00   93.75     0.00   0.07   0.00   0.00   
Failing               0.0       0.00    0.00    66.67   0.00   0.01   0.00   
Fair                 25.0       0.00    0.00     0.00  32.87   3.72  27.34   
Good                  0.0      20.85    0.00     0.00  18.37  61.83   8.63   
Poor                  0.0       0.00    0.00     0.00   5.93   0.46  20.14   
Satisfactory         25.0       0.56    0.00    33.33  41.36  29.99  38.85   
Serious              25.0       0.00    0.00     0.00   0.66   0.11   3.60   
Very Good             0.0      35.77    0.00     0.00   0.59   3.85   0.00   

Superstr_rating  Satisfactory  Serious  Very Good  
Substr_rating                                      
Critical                 0.12     0.00       0.00  
Excellent                0.00     0.00       0.04  
Failed                   0.00     0.00       0.00  
Failing                  0.01     0.00       0.00  
Fair                    13.81    36.36       0.60  
Good                    29.32    18.18      65.77  
Poor                     1.96    13.64       0.03  
Satisfactory            54.03    27.27       8.38  
Serious                  0.29     4.55       0.00  
Very Good                0.46     0.00      25.18

Here we are instead normalizing by index which means to normalize by row. So 93.75% of failed substructures also have a failed superstructure and 6.25% of failed substructures have a fair superstructure. 

In [11]:
sub_sup3 = pd.crosstab(bridges.Substr_rating, bridges.Superstr_rating, normalize='index')
sub_sup3 = sub_sup3.round(4)*100
sub_sup3

Superstr_rating  Critical  Excellent  Failed  Failing   Fair   Good   Poor  \
Substr_rating                                                                
Critical             5.00       0.00    5.00     0.00  10.00  15.00  10.00   
Excellent            0.00      96.20    0.00     0.00   0.00   1.90   0.00   
Failed               0.00       0.00   93.75     0.00   6.25   0.00   0.00   
Failing              0.00       0.00    0.00    50.00   0.00  25.00   0.00   
Fair                 0.04       0.00    0.00     0.00  18.91  24.29   1.60   
Good                 0.00       0.42    0.00     0.00   1.42  54.17   0.07   
Poor                 0.00       0.00    0.00     0.00  22.25  19.78   7.69   
Satisfactory         0.01       0.02    0.00     0.01   5.21  42.91   0.50   
Serious              1.69       0.00    0.00     0.00  15.25  28.81   8.47   
Very Good            0.00       4.65    0.00     0.00   0.29  21.85   0.00   

Superstr_rating  Satisfactory  Serious  Very Good  
Substr_rating                                      
Critical                55.00     0.00       0.00  
Excellent                0.00     0.00       1.90  
Failed                   0.00     0.00       0.00  
Failing                 25.00     0.00       0.00  
Fair                    52.84     0.34       1.98  
Good                    15.04     0.02      28.86  
Poor                    48.90     0.82       0.55  
Satisfactory            45.28     0.06       6.01  
Serious                 44.07     1.69       0.00  
Very Good                1.54     0.00      71.67